### ISSUE TITLE
Connection to gateway not secure (on android) due to TLS 1.0 and deprecated ciphers

### ISSUE THREAD

**Issue by d-bugger** [openmotics/gateway#62]
Since today, chrome on Android marks connection to the gateway (which has a valid certificate) as insecure.
I assume a recent update of the chrome app is causing this.

The reason for marking the connection as insecure is probably due to the weak ciphers.
I see TLS_RSA_WITH_DES_CBC_SHA and TLS_RSA_WITH_3DES_EDE_CBC_SHA in the list, both are <128bits and should be disabled.
IDEA (TLS_RSA_WITH_IDEA_CBC_SHA) also seems vulnerable to the SWEET32 attack
RC4 should no longer be used (deprecated by RFC 7465)
The others look OK.

Please also disable SSLv3 (vulnerable to POODLE attack) ... and actually SSLv2 as well. Only TLS should be activated.

I found this useful resource (see answer): https://stackoverflow.com/questions/34832932/disable-weak-ciphers-with-cherrypy-python-2
with the SSL context, protocol options and ciphers can be set.

Use https://mozilla.github.io/server-side-tls/ssl-config-generator/ to generate a list of strong ciphers supported for the installed openssl version (1.0.0j is no longer "supported" by the way, so maybe a good idea to upgrade to >=1.0.2 while in "security mode" :)
![screenshot_20180309-211754](https://user-images.githubusercontent.com/10128031/37231813-e0852fde-23ec-11e8-91ee-410b49b99951.png)

**Comment by d-bugger** [comment]
Also check the output of testssl.sh against the gateway for insights on the active SSL/TLS vulnerabilities

**Comment by khenderick** [comment]
Thanks for your remarks. However, you mention that the installed certificate is valid which is non-standard as we ship our gateways with self-signed certificates so all browsers give a security warning on which a user can accept the certificate. Did you install a non self-signed certificate yourself (e.g. from let's encrypt)?

**Comment by d-bugger** [comment]
Correct, I'm not using the self-signed certificate ... but weak ciphers and SSLv3 should be disabled in a future release.
I should probably test if I have the same behavior with the self-signed certificate, but I would expect a yes since "accepting" the self-signed certificate is about trusting the identity of the remote party. The "obsolete cypher suite" message is more about the quality of the encryption.

**Comment by khenderick** [comment]
Interesting, I have a let's encrypt certificate myself and I don't see the warnings at this moment. But indeed, I certainly agree that we better have a good look at getting the security at a higher level.

**Comment by d-bugger** [comment]
Just checked and I no longer get an error as well. I see I did get a Chrome update overnight ... checking the security status of the connection, I still see the same text, about the weak ciphers.
So I guess this was triggered by a bug in Chrome.

I suggest to mark this issue as a lower priority security improvement ... I'll try to help out and see if I can fix the cherrypy config to use a only strong ciphers.
Will have to get my git skill to an acceptable level though :)

Which upstream distribution do you use for the BB? Would be great if there's a more recent openssl package available so we can enable TLS1.1 and TLS1.2 as well

**Comment by woutercoppens** [comment]
I'm currently rewriting the HomeAssistant integration for OpenMotics. The new integration is based on httpx.
HTTPX only allows TLS1.2 and up by default.

My gateway is running version 2.20.5 and it still has TLS1.0.

` nmap --script ssl-enum-ciphers -p 443 192.168.0.131
Starting Nmap 7.80 ( https://nmap.org ) at 2021-11-21 11:17 CET
Nmap scan report for 192.168.0.131
Host is up (0.052s latency).

PORT    STATE SERVICE
443/tcp open  https
| ssl-enum-ciphers:
|   SSLv3:
|     ciphers:
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_DES_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_IDEA_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_RC4_128_MD5 (rsa 2048) - C
|       TLS_RSA_WITH_RC4_128_SHA (rsa 2048) - C
|       TLS_RSA_WITH_SEED_CBC_SHA (rsa 2048) - A
|     compressors:
|       NULL
|     cipher preference: client
|     warnings:
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|       64-bit block cipher DES vulnerable to SWEET32 attack
|       64-bit block cipher IDEA vulnerable to SWEET32 attack
|       Broken cipher RC4 is deprecated by RFC 7465
|       CBC-mode cipher in SSLv3 (CVE-2014-3566)
|       Ciphersuite uses MD5 for message integrity
|       Forward Secrecy not supported by any cipher
|       Weak certificate signature: SHA1
|   TLSv1.0:
|     ciphers:
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_CAMELLIA_256_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_DES_CBC_SHA (rsa 2048) - C
|       TLS_RSA_WITH_IDEA_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_RC4_128_MD5 (rsa 2048) - C
|       TLS_RSA_WITH_RC4_128_SHA (rsa 2048) - C
|       TLS_RSA_WITH_SEED_CBC_SHA (rsa 2048) - A
|     compressors:
|       NULL
|     cipher preference: client
|     warnings:
|       64-bit block cipher 3DES vulnerable to SWEET32 attack
|       64-bit block cipher DES vulnerable to SWEET32 attack
|       64-bit block cipher IDEA vulnerable to SWEET32 attack
|       Broken cipher RC4 is deprecated by RFC 7465
|       Ciphersuite uses MD5 for message integrity
|       Forward Secrecy not supported by any cipher
|       Weak certificate signature: SHA1
|_  least strength: C`

My code works perfectly with cloud.openmotics.com as it support TLS1.2 and higher.

`Nmap scan report for cloud.openmotics.com (108.128.34.246)
Host is up (0.037s latency).
rDNS record for 108.128.34.246: ec2-108-128-34-246.eu-west-1.compute.amazonaws.com

PORT    STATE SERVICE
443/tcp open  https
| ssl-enum-ciphers:
|   TLSv1.0:
|     ciphers:
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (ecdh_x25519) - A
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA (dh 2048) - A
|     compressors:
|       NULL
|     cipher preference: server
|     warnings:
|       Key exchange (dh 2048) of lower strength than certificate key
|       Key exchange (ecdh_x25519) of lower strength than certificate key
|   TLSv1.1:
|     ciphers:
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (ecdh_x25519) - A
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA (dh 2048) - A
|     compressors:
|       NULL
|     cipher preference: server
|     warnings:
|       Key exchange (dh 2048) of lower strength than certificate key
|       Key exchange (ecdh_x25519) of lower strength than certificate key
|   TLSv1.2:
|     ciphers:
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (ecdh_x25519) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 (ecdh_x25519) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (ecdh_x25519) - A
|       TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_CCM_8 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_CCM (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 (dh 2048) - A
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA (dh 2048) - A
|     compressors:
|       NULL
|     cipher preference: server
|     warnings:
|       Key exchange (dh 2048) of lower strength than certificate key
|       Key exchange (ecdh_x25519) of lower strength than certificate key
|_  least strength: A`

It is a feature requested by other users as the local gateway has faster responses and it doesn't depend on internet acess.
